ai/recommender/average_recommender.py

```diff
@@ -28,16 +28,61 @@ class AverageRecommender(BaseRecommender):
 
     """
 
-    def __init__(self, ml_type='classifier', metric=None, ml_p=None):
-        """Initialize recommendation system."""
-        super().__init__(ml_type, metric, ml_p)
+    def __init__(self, 
```

Comment:
In one of my unit tests in scikit-learn API branch, loading a picked AverageRecommender cannot generate recommendations. Please check it.

Reply:
```python
#!/usr/bin/env python
# coding: utf-8

# In[1]:


from ai.recommender.average_recommender import AverageRecommender
from ai.knowledgebase_utils import load_knowledgebase
from sklearn.model_selection import cross_val_score, ParameterGrid
import pandas as pd
import json
ml_p_file= "docker/dbmongo/files/projects.json"
kb = "data/knowledgebases/sklearn-benchmark-data-knowledgebase-r6.tsv.gz"
class_metafeatures="data/knowledgebases/pmlb_classification_metafeatures.csv.gz"

# get all ml_p
algorithms = json.load(open(ml_p_file, encoding="utf8"))
# exclude algorithm in different mode
algorithms = [a for a in algorithms if a['category'] == "classification"]

result = [] # returned value
good_def = True # checks that json for ML is in good form

def valid_combo(combo, bad_combos):
    """Checks if parameter combination is valid."""
    for bad_combo in bad_combos:
        bc = {}
        for b in bad_combo:
            bc.update(b)
        bad = True
        for k, v in bc.items():
            if combo[k] != v:
                bad = False
    return not bad

for i,x in enumerate(algorithms):
    hyperparams = x['schema'].keys()
    hyperparam_dict = {}
    # get a dictionary of hyperparameters and their values
    for h in hyperparams:
        if 'ui' in x['schema'][h]:
            if 'values' in x['schema'][h]['ui']:
                hyperparam_dict[h] = x['schema'][h]['ui']['values']
            else:
                hyperparam_dict[h] = x['schema'][h]['ui']['choices']
        else:
            good_def = False
    if good_def:
        all_hyperparam_combos = list(ParameterGrid(hyperparam_dict))
        
        for ahc in all_hyperparam_combos:
            if 'invalidParameterCombinations' in x.keys():
                if not valid_combo(ahc, x['invalidParameterCombinations']):
                    continue
            result.append({'algorithm':x['name'],
                           'category':x['category'],
                           'parameters':ahc,
                           'alg_name':x['name']})
    else:
        logger.error('warning: ' + str(x['name']) + 'was skipped')
    good_def = True

all_ml_p = pd.DataFrame(result)


args = {'metric': 'accuracy', 'ml_p': all_ml_p}


# In[4]:


# load kb
kb = load_knowledgebase(
                            resultsFiles=[kb],
                            metafeaturesFiles=[class_metafeatures]
                            )

# replace algorithm names with their ids
#self.ml_name_to_id = {v:k for k,v in self.ml_id_to_name.items()}
#kb['resultsData']['algorithm'] = kb['resultsData']['algorithm'].apply(
#                                  lambda x: self.ml_name_to_id[x])

all_df_mf = kb['metafeaturesData'].set_index('_id', drop=False)

# all_df_mf = pd.DataFrame.from_records(metafeatures).transpose()
# keep only metafeatures with results
df = all_df_mf.loc[kb['resultsData']["classification"]['_id'].unique()]


# In[2]:


rec_engines = AverageRecommender(**args)
rec_engines.update(kb['resultsData']["classification"], df, source='knowledgebase')


# In[3]:


from ai.metalearning.dataset_describe import Dataset
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import numpy as np
from ai.metalearning.get_metafeatures import generate_metafeatures
iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data.astype(np.float64),
    iris.target.astype(np.float64), train_size=0.75, test_size=0.25, random_state=42)
columns = ["Feature_{}".format(i) for i in range(X_train.shape[1])]
df = pd.DataFrame(X_train, columns=columns)
df['pennai_target'] = y_train
df.head()


# In[4]:


dataset = Dataset(df=df,
                          dependent_col="pennai_target",
                          prediction_type="classification"
                          )
datasetId = dataset.m_data_hash()
meta_features = generate_metafeatures(dataset)


# In[5]:


datasetId


# In[6]:


ml, p, ai_scores = rec_engines.recommend(
            dataset_id=datasetId,
            n_recs=2,
            dataset_mf=meta_features)
recommendations = []
for alg, params, score in zip(ml, p, ai_scores):
        recommendations.append({'dataset_id': datasetId,
                                'algorithm': alg,
                                'parameters': params,
                                'ai_score': score,
                                })
print(recommendations) # this works


# In[7]:


rec_engines.save("pickled_average_recommender.pkl.gz")


# In[8]:


pickle_args = args
pickle_args['serialized_rec_filename'] = "pickled_average_recommender.pkl.gz"
pickle_args['serialized_rec_directory'] = "./"
pickle_args['load_serialized_rec'] = "always"
pickle_args['knowledgebase_results'] = kb['resultsData']["classification"]

pickle_rec_engines = AverageRecommender(**pickle_args)


# In[9]:


ml, p, ai_scores = pickle_rec_engines.recommend(
            dataset_id=datasetId,
            n_recs=2,
            dataset_mf=meta_features)
recommendations = []
for alg, params, score in zip(ml, p, ai_scores):
        recommendations.append({'dataset_id': datasetId,
                                'algorithm': alg,
                                'parameters': params,
                                'ai_score': score,
                                })
print(recommendations) # this does not have any recommendations

```
Above is a demo for reproducing this issue on pickle_recommenders_241 branch

Reply:
fixed here 980fdbf